src/core/mount.c

```diff
@@ -2458,6 +2461,58 @@ static const char* const mount_result_table[_MOUNT_RESULT_MAX] = {
 
 DEFINE_STRING_TABLE_LOOKUP(mount_result, MountResult);
 
+static bool mount_can_freeze(Unit *u) {
+        Mount *m = MOUNT(u);
+        const char *fstype;
+
+        assert(u);
+        assert(m);
+
+        fstype = mount_get_fstype(m);
+        if (!fstype)
+                return false;
+
+        return fstype_can_freeze_thaw(fstype);
+}
+
+static int mount_freezer_action(Unit *u, FreezerAction action) {
+        _cleanup_close_ int fd = -EBADF;
+        Mount *m = MOUNT(u);
+        FreezerState next, target;
+        int r;
+
+        assert(u);
+        assert(m);
+        assert(IN_SET(action, FREEZER_FREEZE, FREEZER_PARENT_FREEZE,
+                              FREEZER_THAW, FREEZER_PARENT_THAW));
+
+        unit_next_freezer_state(u, action, &next, &target);
+        next = freezer_state_finish(next); /* We're completely sync */
+
+        log_unit_debug(u, "Mount freezer state was %s, now %s.",
+                       freezer_state_to_string(u->freezer_state),
+                       freezer_state_to_string(next));
+
+        fd = open(m->where, O_RDONLY|O_NOCTTY);
+        if (fd < 0)
+                return -errno;
+
+        if (target == FREEZER_FROZEN) {
+                r = RET_NERRNO(ioctl(fd, FIFREEZE, 0));
+                if (r == -EBUSY) /* Already frozen by us */
+                        return 0;
+        } else {
+                r = RET_NERRNO(ioctl(fd, FITHAW, 0));
+                if (r == -EINVAL) /* Not frozen by us */
+                        return 0;
```

Comment:
logging is sometimes a bit wonky in pid 1, and needs some clean-ups, but i think we should do log_unit_warning() here for unexpected freeze/thaw errors at least.

Reply:
I did log_unit_error instead, since we don't ignore the exit value and do report failure to our caller